metrics.nim

```diff
@@ -542,6 +553,135 @@ template time*(summary: Summary, body: untyped) =
+                    buckets: openArray[float64] = defaultHistogramBuckets): Histogram =
+    validateName(name)
+    validateLabels(labels, invalidLabelNames = ["le"])
+    var bucketsSeq = @buckets
+    if bucketsSeq.len > 0 and bucketsSeq[^1] != Inf:
+      bucketsSeq.add(Inf)
+    if bucketsSeq.len < 2:
+      raise newException(ValueError, "Invalid buckets list: '" & $bucketsSeq & "'. At least 2 required.")
+    if not bucketsSeq.isSorted(system.cmp[float64]):
+      raise newException(ValueError, "Invalid buckets list: '" & $bucketsSeq & "'. Must be sorted.")
+    result = Histogram(name: name,
+                    help: help,
+                    typ: "histogram",
+                    labels: @labels,
+                    metrics: initOrderedTable[Labels, seq[Metric]](),
+                    creationThreadId: getThreadId(),
+                    buckets: bucketsSeq)
+    if labels.len == 0:
+      result.metrics[@labels] = newHistogramMetrics(name, labels, labels, bucketsSeq)
+    result.register(registry)
+
+template declareHistogram*(identifier: untyped,
+                        help: static string,
+                        labels: LabelsParam = @[],
+                        registry = defaultRegistry,
+                        buckets: openArray[float64] = defaultHistogramBuckets) {.dirty.} =
+  when defined(metrics):
+    var identifier = newHistogram(astToStr(identifier), help, labels, registry, buckets)
+  else:
+    type identifier = IgnoredCollector
+
+template declarePublicHistogram*(identifier: untyped,
+                               help: static string,
+                               labels: LabelsParam = @[],
+                               registry = defaultRegistry,
+                               buckets: openArray[float64] = defaultHistogramBuckets) {.dirty.} =
+  when defined(metrics):
+    var identifier* = newHistogram(astToStr(identifier), help, labels, registry, buckets)
+  else:
+    type identifier* = IgnoredCollector
+
+when defined(metrics):
+  proc histogram*(name: static string): Histogram =
+    var res {.global.} = newHistogram(name, "") # lifted line
+    return res
+else:
+  template histogram*(name: static string): untyped =
+    IgnoredCollector
+
+proc observeHistogram(histogram: Histogram, amount: int64|float64, labelValues: LabelsParam = @[]) =
+  when defined(metrics):
+    var timestamp = getTime().toMilliseconds()
+
+    let labelValuesCopy = validateHistogramLabelValues(histogram, labelValues)
+
+    atomicAdd(histogram.metrics[labelValuesCopy][0].value.addr, amount.float64) # _sum
+    atomicStore(cast[ptr int64](histogram.metrics[labelValuesCopy][0].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+    atomicAdd(histogram.metrics[labelValuesCopy][1].value.addr, 1.float64) # _count
+    atomicStore(cast[ptr int64](histogram.metrics[labelValuesCopy][1].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+    for i, bucket in histogram.buckets:
+      if amount.float64 <= bucket:
```

Comment:
Perhaps the buckets can be defined by a formula instead of listing the boundaries as arbitrary numbers. Then it would be possible to apply the formula here as well and get the correct bucket without searching. I guess it's also possible to define the case of arbitrary boundaries as a special formula that uses table lookups.

Reply:
Why complicate things before we get some data on whether this simple approach is a problem or not in practice?